### PR TITLE
Mjcschool task3

### PR DESCRIPTION
**Pull request by zum1k**
Complete task 3(module #3. REST API Advanced)

### CODE REVIEW

File: src/main/generated/com/epam/esm/service/mapper/order/OrderMapperImpl.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+        List<OrderDto> list = new ArrayList<OrderDto>( orders.size() );
+        for ( Order order : orders ) {
+            list.add( toDto( order ) );
+        }
+
+        return list;
+    }
+
+    private long orderUserUserId(Order order) {
```
Comment (n1k1taM) [inline src/main/generated/com/epam/esm/service/mapper/order/OrderMapperImpl.java:78]: The method name is not clear.

File: src/main/java/com/epam/esm/repository/specification/SpecificationCreator.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.ArrayList;
+import java.util.List;
+@Service
+public class SpecificationCreator {
+    public List<CriteriaSpecification<GiftCertificate>> createSpecifications(RequestParametersDto dto){
+        List<CriteriaSpecification<GiftCertificate>> specifications = new ArrayList<>();
+        if(dto.getTagName()!=null){
+            specifications.add(new CertificatesByTagNameCriteriaSpecification(dto.getTagName()));
+        }
+        if(dto.getPartName()!=null){
```
Comment (n1k1taM) [inline src/main/java/com/epam/esm/repository/specification/SpecificationCreator.java:17]: You can use `Objects.nonNull(dto.getPartName())`

File: src/test/java/com/epam/esm/repository/tag/TagRepositoryImplTest.java [hunk 3]
```diff
@@ -1,64 +1,104 @@
...
+  void removeTag_ShouldReturn_True_Test() {
+    long expectedId = 9;
+    long actualResult = tagRepository.remove(expectedId).get().getTagId();
+    ;
+    Assertions.assertEquals(expectedId, actualResult);
+  }
+
+  @Test
+  @Transactional
+  void findByName_ShouldReturn_True_Test() {
```
Comment (n1k1taM) [inline src/test/java/com/epam/esm/repository/tag/TagRepositoryImplTest.java:45]: Where are negative test cases?

File: src/test/java/com/epam/esm/repository/user/UserRepositoryImplTest.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+@SpringBootTest
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+class UserRepositoryImplTest {
+  private final UserRepository repository;
+
+  @Test
+  @Transactional
+  void find_ShouldReturn_User_True() {
+    long expectedId = 4;
+    long actualId = repository.find(expectedId).get().getUserId();
```
Comment (n1k1taM) [inline src/test/java/com/epam/esm/repository/user/UserRepositoryImplTest.java]: Can we get `NullPointerException` here?

File: src/test/java/com/epam/esm/service/certificate/GiftServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+
+    Mockito.when(repository.add(certificate)).thenReturn(Optional.of(certificate));
+    Mockito.when(mapper.toEntity(certificateDto)).thenReturn(certificate);
+    Mockito.when(mapper.toDto(certificate)).thenReturn(certificateDto);
+    Mockito.when(tagService.addTagIfNotExist(Mockito.any(TagDto.class))).thenReturn(tagDto);
+    Mockito.when(tagMapper.toEntity(tagDto)).thenReturn(tag);
+
+    String actualName = service.add(certificateDto).getName();
+
+    Assertions.assertEquals(expectedName, actualName);
```
Comment (n1k1taM) [inline src/test/java/com/epam/esm/service/certificate/GiftServiceImplTest.java:77]: I guess that we we have to use `verify` method to control that mock was called

File: src/test/java/com/epam/esm/service/certificate/GiftServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+    Assertions.assertThrows(
+        EntityNotFoundException.class,
+        () -> {
+          service.findById(expectedCertificateID);
+        });
+    Mockito.verify(repository).findById(Mockito.eq(expectedCertificateID));
+  }
+
+  @Test
+  void count_ShouldReturn_Ten_Test() {
```
Comment (n1k1taM) [inline src/test/java/com/epam/esm/service/certificate/GiftServiceImplTest.java]: It's a bad practice to use number as part of testing method's name.

File: src/main/generated/com/epam/esm/service/mapper/order/OrderMapperImpl.java (hunk 1 above)
Comment (zum1k) [reply src/main/generated/com/epam/esm/service/mapper/order/OrderMapperImpl.java:78]: This is auto-generated methods by "mapstruct" library

File: src/test/java/com/epam/esm/repository/tag/TagRepositoryImplTest.java (hunk 3 above)
Comment (zum1k) [reply src/test/java/com/epam/esm/repository/tag/TagRepositoryImplTest.java:45]: Negative case "empty()" is in TagServiceImplTest.java

File: src/test/java/com/epam/esm/repository/user/UserRepositoryImplTest.java (hunk 4 above)
Comment (zum1k) [reply src/test/java/com/epam/esm/repository/user/UserRepositoryImplTest.java]: Fixed( add empty() case in RepositoryUser)

File: src/test/java/com/epam/esm/service/certificate/GiftServiceImplTest.java (hunk 6 above)
Comment (zum1k) [reply src/test/java/com/epam/esm/service/certificate/GiftServiceImplTest.java]: Fixed to count_BySpecifications_ShouldReturn_10Pages_Test . There number(10) is a part of expected behavior of test method.

File: src/test/java/com/epam/esm/service/certificate/GiftServiceImplTest.java (hunk 5 above)
Comment (zum1k) [reply src/test/java/com/epam/esm/service/certificate/GiftServiceImplTest.java:77]: Fixed

**Pull request merged**